Bio/bgzf.py

```diff
@@ -633,59 +633,55 @@ def read(self, size=-1):
         """Read method for the BGZF module."""
         if size < 0:
             raise NotImplementedError("Don't be greedy, that could be massive!")
-        elif size == 0:
-            if self._text:
-                return ""
-            else:
-                return b""
-        elif self._within_block_offset + size <= len(self._buffer):
-            # This may leave us right at the end of a block
-            # (lazy loading, don't load the next block unless we have too)
-            data = self._buffer[self._within_block_offset:self._within_block_offset + size]
-            self._within_block_offset += size
-            assert data  # Must be at least 1 byte
-            return data
-        else:
-            data = self._buffer[self._within_block_offset:]
-            size -= len(data)
-            self._load_block()  # will reset offsets
-            # TODO - Test with corner case of an empty block followed by
-            # a non-empty block
-            if not self._buffer:
-                return data  # EOF
-            elif size:
-                # TODO - Avoid recursion
-                return data + self.read(size)
+
+        result = "" if self._text else b""
+        while size and self._buffer:
+            if self._within_block_offset + size <= len(self._buffer):
+                # This may leave us right at the end of a block
+                # (lazy loading, don't load the next block unless we have too)
+                data = self._buffer[self._within_block_offset:self._within_block_offset + size]
+                self._within_block_offset += size
+                assert data  # Must be at least 1 byte
```

Comment:
In the future it might be nice to have an `Exception` here in case the assertion is false. Something like `ValueError("Expected data at byte %d." % self.tell())`. For now I think this looks fine.

Reply:
Fair point, yes.